File: src/main/java/de/crazydev22/resourcesftp/FtpHost.java
package de.crazydev22.resourcesftp;

import org.gradle.api.credentials.PasswordCredentials;
import java.net.URI;
import java.util.Objects;

public class FtpHost {
    private final String hostname;
    private final int port;
    private final String username;
    private final String password;
    private final Security security;

    public FtpHost(URI uri, PasswordCredentials credentials) {
        this.hostname = uri.getHost();
        this.port = uri.getPort();
        this.username = credentials != null ? credentials.getUsername() : null;
        this.password = credentials != null ? credentials.getPassword() : null;
        this.security = Security.fromURI(uri);
    }

    public String getHostname() {
        return hostname;
    }

    public int getPort() {
        return port;
    }

    public String getUsername() {
        return username;
    }

    public String getPassword() {
        return password;
    }

    public Security getSecurity() {
        return security;
    }

    public boolean shouldLogin() {
        return username != null;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        FtpHost ftpHost = (FtpHost) o;
        return port == ftpHost.port &&
                Objects.equals(hostname, ftpHost.hostname) &&
                Objects.equals(username, ftpHost.username) &&
                Objects.equals(password, ftpHost.password);
    }

    @Override
    public int hashCode() {
        return Objects.hash(hostname, port, username, password);
    }

    @Override
    public String toString() {
        return String.format("%s:%d (Username: %s)", hostname, port, username);
    }
}

File: src/main/kotlin/de/crazydev22/resourcesftp/FtpHost.kt
package de.crazydev22.resourcesftp

import org.gradle.api.credentials.PasswordCredentials
import java.net.URI
import java.util.*

class FtpHost(uri: URI, credentials: PasswordCredentials?) {
    val hostname: String = uri.host
    val port: Int = uri.port
    val username: String? = credentials?.username
    val password: String? = credentials?.password
    val ssl: Boolean = uri.scheme == "ftps"

    fun shouldLogin(): Boolean {
        return username != null && password != null
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val ftpHost = other as FtpHost
        return port == ftpHost.port && hostname == ftpHost.hostname && username == ftpHost.username && password == ftpHost.password
    }

    override fun hashCode(): Int {
        return Objects.hash(hostname, port, username, password)
    }

    override fun toString(): String {
        return String.format("%s:%d (Username: %s)", hostname, port, username)
    }
}
